### PR TITLE
Rin/azure vulns

### PR DESCRIPTION
**Pull request by rin-skylight**
# DEVOPS PULL REQUEST

## Related Issue

- This resolves a currently unlisted issue raised by the CDC to close access to our existing app services and function apps using their default URLs. All traffic will instead route through the application gateway.
- Addresses several issues raised as part of #4245.

## Changes Proposed

- Restricts API traffic to application gateway.
- Restricts Metabase traffic to application gateway.
- Reworks Makefile and CI jobs to use new URL flow.
- Enforces https and TLS usage across relevant services.

## Additional Information

- Existing URLs that are used for testing and targeting no longer function as a result of this change. Please note the following paths that should be used moving forward:
  - Staging endpoints: https://ENV.simplereport.gov/staging/api (accesses all API functions within the staging slot)
  - Metabase: https://ENV.simplereport.gov/metabase
- The `Makefile` provides relevant examples of how to access the various actuator endpoints.

## Testing

- A full deployment should be tested on an environment that also has metabase. Any of the dev-level environments should work, with the exception of Dev3, as that environment has been migrated.
- A full smoke test should be completed, including interactions with Okta.

### Infrastructure
- [x] Consult the results of the `terraform-plan` job inside the "Terraform Checks" workflow run for this PR. Confirm that there are no unexpected changes!

### Security
- [x] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [x] Any dependencies introduced have been vetted and discussed

### Cloud
- [x] Oncall has been notified if this change is going in after-hours
- [x] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

### Documentation
- [x] Any changes to the startup configuration have been documented in the README

### CODE REVIEW

File: ops/Makefile [hunk 1]
```diff
@@ -59,55 +63,103 @@ promote-%: .be-logged-in .valid-env-%
...
+	if [[ "$*" == "prod" ]] ; then \
+	  [[ "$(RELEASE_TAG)"  == "`$(CURL) $(PROD_INFO_ENDPOINT) | jq -r .deploy.tag`" ]] ; \
+	else \
+	  [[ "$(RELEASE_TAG)"  == "`$(CURL) $(INFO_ENDPOINT) | jq -r .deploy.tag`" ]] ; \
+	fi;
 
 check-%-slot-readiness: .valid-env-%
 	@echo Checking the readiness probe for $*/$(SOURCE_SLOT)
-	$(SLOT_READY)
+	@if [[ "$*" == "prod" ]] ; then \
```
Comment (emmastephenson) [inline ops/Makefile:150]: Curious why we want to suppress the echo of `if` only for this conditional and the one below? (I don't interact with Makefiles very often, so apologies if this is a silly question!)

**Review by emmastephenson — COMMENTED** [review]
Code changes LGTM, but holding off on approval until I have a chance to test this.

@rin-skylight, do you want us to set up a mini-bugbash to verify that everything is still WAI after these changes? I'm thinking a session where we 
1. Deploy to a remote env and validate that the deployment works
2. Hammer on Metabase for a bit
3. Verify that tests are published to ReportStream correctly, since the function app changed

File: ops/services/app_gateway/main.tf [hunk 2]
```diff
@@ -235,12 +314,58 @@ resource "azurerm_application_gateway" "load_balancer" {
...
+    }
+  }
+
   rewrite_rule_set {
     name = "simple-report-routing"
 
     rewrite_rule {
       name          = "api-wildcard"
-      rule_sequence = 100
+      rule_sequence = 101
```
Comment (emyl3) [inline ops/services/app_gateway/main.tf:368]: is there a link you have handy to what `rule_sequence` being set to `101` means as opposed to `100`? 🔍

**Review by emyl3 — COMMENTED** [review]
Thank you for all your work on this, Rin! (Most of this went over my head 😅 so I won't approve... but I had one question for you!) 

I do like Emma's idea of doing a mini bug bash to test these changes if you think that's a good idea as well :)

File: ops/Makefile (hunk 1 above)
Comment (rin-skylight) [reply ops/Makefile]: I appreciate you calling this out! That was my error on copy-paste. Fixed!

**Comment by rin-skylight** [comment]
> Code changes LGTM, but holding off on approval until I have a chance to test this.
> 
> @rin-skylight, do you want us to set up a mini-bugbash to verify that everything is still WAI after these changes? I'm thinking a session where we
> 
> 1. Deploy to a remote env and validate that the deployment works
> 2. Hammer on Metabase for a bit
> 3. Verify that tests are published to ReportStream correctly, since the function app changed

@emmastephenson I would love to have a bugbash for this! I want to make sure these changes are as airtight as possible. `dev3` has all of of this on it already, and is available for the team to test at any time.

File: ops/Makefile [hunk 3]
```diff
@@ -59,55 +63,103 @@ promote-%: .be-logged-in .valid-env-%
 
 wait-for-%-slot-commit: .valid-env-%
 	@echo "Starting wait for commit $(DEPLOYED_COMMIT) in $(SOURCE_SLOT) slot of $* at `date`"
-	until  [[ "$(DEPLOYED_COMMIT)"  == "`$(CURL) $(SLOT_INFO_ENDPOINT) | jq -r .git.commit.id`" ]] ; do sleep 5; done;
+	if [[ "$*" == "prod" ]] ; then \
+	  until  [[ "$(DEPLOYED_COMMIT)"  == "`$(CURL) $(PROD_SLOT_INFO_ENDPOINT) | jq -r .git.commit.id`" ]] ; do sleep 5; done; \
+	else \
+	  until  [[ "$(DEPLOYED_COMMIT)"  == "`$(CURL) $(SLOT_INFO_ENDPOINT) | jq -r .git.commit.id`" ]] ; do sleep 5; done; \
```
Comment (alismx) [inline ops/Makefile:69]: Question: Would it be possible to check for `prod` during the variable assignment where we form the endpoints? (lines 14-17 and 19-22) If so, I think can cut out these repetitive `if/else` checks.

Comment (rin-skylight) [reply ops/Makefile:79]: Noting here for the record: We discussed this in our DevOps standups this week, and have agreed that this isn't a showstopper for this PR. We may end up cleaning this up at a later date.

**Comment by rin-skylight** [comment]
This PR is now in its final state, and is ready for official review.

**Review by emmastephenson — APPROVED** [review]
LGTM! Tested on `dev4` and everything that was identified in the bug bash looks to be fixed. 

Thanks for all your work on this, Rin! 🙇

**Review by emyl3 — APPROVED** [review]
LGTM! Went through all the bug bash items and were able to confirm they work as intended now. Thank you, Rin!!

**Pull request merged**